offload/new-api/src/offload_impl.cpp

```diff
@@ -0,0 +1,190 @@
+//===- offload_impl.cpp - Implementation of the new LLVM/Offload API ------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This contains the definitions of the new LLVM/Offload API entry points. See
+// new-api/API/README.md for more information.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PluginManager.h"
+#include "helpers.hpp"
+#include "llvm/Support/FormatVariadic.h"
+#include <offload_api.h>
+
+#include <mutex>
+
+using namespace llvm;
+using namespace llvm::omp::target::plugin;
+
+// Handle type definitions. Ideally these would be 1:1 with the plugins
+struct offload_device_handle_t_ {
+  int DeviceNum;
+  GenericDeviceTy &Device;
+};
+
+struct offload_platform_handle_t_ {
+  std::unique_ptr<GenericPluginTy> Plugin;
+  std::vector<offload_device_handle_t_> Devices;
```

Comment:
These should be LLVM small vectors, since they have small object optimizations.

Reply:
Also, similar for anything `std::` there's probably something like `llvm::find_if` that's easier to use. Can you also run clang-tidy on the whole file to make sure everything's formatted with the right names?